### PR TITLE
Make Strategy an exported class in passport

### PR DESCRIPTION
**Pull request by mohsen1**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by mohsen1** [comment]
It's an actual class:


```
> const passport = require('passport')
undefined
> class MyStrategy extends passport.Strategy {}
undefined
> const m = new MyStrategy()
undefined
> m.authenticate()
Error: Strategy#authenticate must be overridden by subclass
    at MyStrategy.Strategy.authenticate (/Users/mazimi/lyft/expressdrivefe/node_modules/passport-strategy/lib/strategy.js:21:9)
>

```

**Comment by cjbarth** [comment]
Could you please reference some documentation or source code that shows that we should be using `extends` instead of `implements`?

**Comment by mohsen1** [comment]
@cjbarth Here you can see passport is exporting this class:

https://github.com/jaredhanson/passport/blob/df6e2b7938ebd92c26491e47b73b20ee3516f823/lib/index.js#L20

File: types/passport/index.d.ts [hunk 1]
```diff
@@ -73,6 +73,7 @@ declare namespace passport {
     interface PassportStatic extends Authenticator {
         Authenticator: { new(): Authenticator };
         Passport: PassportStatic["Authenticator"];
+        Strategy: { new(): Strategy & StrategyCreatedStatic };
```
Comment (mohsen1) [inline types/passport/index.d.ts]: This is the main change, rest is test and related package updates.

**Review by dsluijk — APPROVED** [review]

**Comment by theigor** [comment]
Guys, I think you broke something. I just got latest and now I am getting this on `tsc`

```
node_modules/@types/passport-strategy/index.d.ts(19,32): error TS2689: Cannot extend an interface 'passport.Strategy'. Did you mean 'implements'?
```

**Comment by cjbarth** [comment]
@theigor I had the same problem. You'll have to pin all your types to the previous version unless you can move your `passport` types to `^0.4.0`. My `passport` types were at `^0.3.0`, and the change to `extend` landed in connection with the `^.0.4.0` series. If you need the exact versions that work, I can get that for you, just let me know.

**Comment by sddddgjd** [comment]
@cjbarth i'm running into the same issue, could you please provide the exact versions that work?

**Comment by mohsen1** [comment]
Can you paste your error message?

**Comment by sddddgjd** [comment]
Same error message:

`node_modules/@types/passport-strategy/index.d.ts(19,32): error TS2689: Cannot extend an interface 'passport.Strategy'. Did you mean 'implements'?`

Worked around it by using a previous versions for passport-local types

**Comment by mohsen1** [comment]
update `@types/passport` and all other `@types/passport-*` packages. It should fix your issue

**Comment by cjbarth** [comment]
@sddddgjd If you use `@types/passport: ^0.3.3`, then you'll have to stick to `@types/passport-http: 0.3.5`, `@types/passport-saml: 0.15.5`, and `@types/passport-strategy: 0.2.32`.

**Comment by Brimill** [comment]
Hey, I ran into some compile errors with the tests and am unsure what causes them.
`passport-tests.ts:13:18 - error TS2339: Property 'fail' does not exist on type 'TestStrategy'.`
`passport-tests.ts:15:18 - error TS2339: Property 'success' does not exist on type 'TestStrategy'.`

**Comment by cjbarth** [comment]
@Brimill You are commenting on a merged PR. It is very unlikely that you'll get much help here. Additionally, the information you've provided above is too little to help you. Please give more information about what you are doing when you get these errors, commands, environment, etc.

**Pull request merged**